k9mail/src/test/java/com/fsck/k9/controller/MessagingControllerTest.java

```diff
@@ -0,0 +1,379 @@
+    private static final int MAXIMUM_SMALL_MESSAGE_SIZE = 1000;
+
+
+    private MessagingController controller;
+    @Mock
+    private Account account;
+    @Mock
+    private AccountStats accountStats;
+    @Mock
+    private MessagingListener listener;
+    @Mock
+    private LocalFolder localFolder;
+    @Mock
+    private Folder remoteFolder;
+    @Mock
+    private LocalStore localStore;
+    @Mock
+    private Store remoteStore;
+    @Mock
+    private NotificationController notificationController;
+    @Captor
+    private ArgumentCaptor<List<Message>> messageListCaptor;
+    @Captor
+    private ArgumentCaptor<FetchProfile> fetchProfileCaptor;
+
+
+    @Before
+    public void setUp() throws MessagingException {
+        MockitoAnnotations.initMocks(this);
+        Context appContext = ShadowApplication.getInstance().getApplicationContext();
+
+        controller = new MessagingController(appContext, notificationController);
+
+        configureAccount();
+        configureLocalStore();
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        controller.stop();
+    }
+
+    @Test
+    public void synchronizeMailboxSynchronous_withOneMessageInRemoteFolder_shouldFinishWithoutError() throws Exception {
+        messageCountInRemoteFolder(1);
+
+        controller.synchronizeMailboxSynchronous(account, FOLDER_NAME, listener, remoteFolder);
+
+        verify(listener).synchronizeMailboxFinished(account, FOLDER_NAME, 1, 0);
+    }
+
+    @Test
+    public void synchronizeMailboxSynchronous_withEmptyRemoteFolder_shouldFinishWithoutError() throws Exception {
+        messageCountInRemoteFolder(0);
+
+        controller.synchronizeMailboxSynchronous(account, FOLDER_NAME, listener, remoteFolder);
+
+        verify(listener).synchronizeMailboxFinished(account, FOLDER_NAME, 0, 0);
+    }
+
+    @Test
+    public void synchronizeMailboxSynchronous_withNegativeMessageCountInRemoteFolder_shouldFinishWithError()
+            throws Exception {
+        messageCountInRemoteFolder(-1);
+
+        controller.synchronizeMailboxSynchronous(account, FOLDER_NAME, listener, remoteFolder);
+
+        verify(listener).synchronizeMailboxFailed(account, FOLDER_NAME,
+                "Exception: Message count -1 for folder Folder");
+    }
+
+    @Test
+    public void synchronizeMailboxSynchronous_withRemoteFolderProvided_shouldNotOpenRemoteFolder() throws Exception {
+        messageCountInRemoteFolder(1);
+
+        controller.synchronizeMailboxSynchronous(account, FOLDER_NAME, listener, remoteFolder);
+
+        verify(remoteFolder, never()).open(Folder.OPEN_MODE_RW);
+
+    }
+
+    @Test
+    public void synchronizeMailboxSynchronous_withNoRemoteFolderProvided_shouldOpenRemoteFolderFromStore() throws Exception {
+        messageCountInRemoteFolder(1);
+        when(account.getRemoteStore()).thenReturn(remoteStore);
+        when(remoteStore.getFolder(FOLDER_NAME)).thenReturn(remoteFolder);
```

Comment:
The above two lines are repeated a lot. Extracting them to a method should clean up the tests quite a bit.